README.md
# [image_pickers](https://github.com/lisen87/image_pickers)

image_pickers支持本地图片多选，本地视频多选，支持将网络图片保存到相册，支持截图保存到相册，支持将网络视频保存到相册，支持预览视频和预览图片功能，支持主题颜色设置

image pickers support multi-selection of local pictures, multi-selection of local videos, support for saving network pictures to albums, support for saving screenshots to albums, support for saving network videos to albums, support for preview videos and preview images, and support for theme color settings

> Supported  Platforms
> * Android
> * iOS

## iOS : platform :ios, '10.0'


## Android: minSdkVersion 19 targetSdkVersion 29 compileSdkVersion 29
![](https://github.com/lisen87/image_pickers/blob/master/screenshots/1.png)


## 裁剪状态下图片selectCount 字段无效，将变为单选 In the cropped state, the selectCount field of the picture is invalid and will become single selection

## How to Use

```yaml
# add this line to your dependencies
image_pickers: ^1.1.0+8
```

```dart
import 'package:image_pickers/image_pickers.dart';
```
```dart


///选择多张图片 Select multiple images
Future<void> selectImages() async {
    List<Media> _listImagePaths = await ImagePickers.pickerPaths(
              galleryMode: GalleryMode.image,
              selectCount: 2,
              showGif: false,
              showCamera: true,
              compressSize: 500,
              uiConfig: UIConfig(uiThemeColor: Color(0xffff0f50)),
              cropConfig: CropConfig(enableCrop: false, width: 2, height: 1));
  }

/// 或者 or
ImagePickers.pickerPaths().then((List medias){
      /// medias 照片路径信息 Photo path information
    });

```
```dart
///选择多个视频 Select multiple videos
Future<void> selectVideos() async {
   List<Media> _listVideoPaths = await ImagePickers.pickerPaths(
          galleryMode: GalleryMode.video,
          selectCount: 5,
        );
  }
```
```dart
///直接打开相机拍摄图片 Open the camera directly to take a picture
ImagePickers.openCamera().then((Media media){
    /// media 包含照片路径信息 Include photo path information
  });

```

```dart
///直接打开相机拍摄视频 Open the camera directly to shoot the video
ImagePickers.openCamera(cameraMimeType: CameraMimeType.video).then((media){
    /// media 包含视频路径信息 Contains video path information
  });

```

```dart
///预览图片 Preview picture
ImagePickers.previewImage(_listImagePaths[index].path);
///预览多张图片 Preview multiple pictures
ImagePickers.previewImagesByMedia(_listImagePaths,index);
///预览多张图片 Preview multiple pictures
ImagePickers.previewImages(paths,index);

///预览视频 Preview video
ImagePickers.previewVideo(_listVideoPaths[index].path);
```
```dart
///保存图片到图库 Save image to gallery
ImagePickers.saveImageToGallery("http://i1.sinaimg.cn/ent/d/2008-06-04/U105P28T3D2048907F326DT20080604225106.jpg");
```

```dart
/// 保存截图图片 ByteData 到图库 Save screenshot image ByteData to gallery
RenderRepaintBoundary boundary = globalKey.currentContext.findRenderObject();
ui.Image image = await boundary.toImage(pixelRatio: 3);
ByteData byteData = await image.toByteData(format: ui.ImageByteFormat.png);
Uint8List data = byteData.buffer.asUint8List();

String dataImagePath = await ImagePickers.saveByteDataImageToGallery(data,);

```


```dart
///保存视频到图库 Save video to gallery
ImagePickers.saveVideoToGallery("http://xxxx/xx/xx.mp4");
```

## iOS
Add the following entry to your `Info.plist` file, located in `/Info.plist` :
`
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsArbitraryLoads</key>
    <true/>
</dict>
<key>NSMicrophoneUsageDescription</key>
<string>...</string>
<key>NSPhotoLibraryAddUsageDescription</key>
<string>...</string>
<key>NSCameraUsageDescription</key>
<string>...</string>
<key>NSPhotoLibraryUsageDescription</key>
<string>...</string>
`


CHANGELOG.md
## 1.1.0+8
* 支持空安全
* null-safe support
## 1.0.8+7
* Android 修复打开相册时，无法录制视频问题
* Android fixes the problem that the video cannot be recorded when opening the gallery
## 1.0.8+6
* iOS 尝试修复最近的各种问题
* iOS Try to fix various recent problems
## 1.0.8+4
* 安卓修复安卓Q出现的问题等
* Android fixes problems with Android Q, etc.
* iOS处理包名错误问题等
* iOS handles package name errors, etc.
## 1.0.8+3
* 更新AFNetworking 到4.0 删除UIWebview ，处理iOS上架失败问题
## 1.0.8+2
* 添加是否显示gif图片字段
* Add whether to display the gif image field
* android 更新权限申请流程，不再一次性申请不相关权限
* android Update the permission application process and no longer apply for irrelevant permissions all at once
* iOS14 修复图片预览无法显示问题/修复无法关闭裁剪问题
* iOS14 fixes the problem that picture preview cannot be displayed/Fix the problem that cropping cannot be turned off

## 1.0.7+6
* iOS 固定ZLPhotoBrowser为3.1.2 
* iOS fixed ZLPhotoBrowser to 3.1.2
## 1.0.7+5
* Android 修复保存屏幕截图到相册时，因文件后缀名造成的崩溃问题。
* Android fixes the crash caused by the file extension when saving screenshots to albums.
## 1.0.7+4
* 修复 iOS 因为图片（right_arrow.png、left_arrow.png、error.png）导致的各种问题
* Fix various problems caused by pictures (right_arrow.png, left_arrow.png, error.png) on ​​iOS
## 1.0.7+3
* 临时修复 MethodChannel.Result 为 null 崩溃问题
* Temporarily fix crash when MethodChannel.Result is null
## 1.0.7+2
* 使用了sdk29，适配Android Q
* Used sdk29, adapted to Android Q
* 安卓修复小米Android Q 问题
* Android fixes Xiaomi Android Q issue
## 1.0.7+1
* 安卓修复安卓Q .gif 图片裁剪后无法播放的问题
* Android fix the problem that Android Q .gif picture cannot be played after cropping
## 1.0.7
* 新增保存图片ByteData数据
* Added Byte Data for saving pictures
* 修复iOS压缩图片异常
* Fix iOS compressed picture abnormal

## 1.0.6+2
* 修复flutter回调异常
* Fix flutter callback exception
* `CorpConfig` 类 改为 `CropConfig` 类，参数中的 `corpConfig` 改为 `cropConfig`
* `CorpConfig` class changed to` CropConfig` class, and `corpConfig` in the parameter was changed to` cropConfig`
* ios修复ios13获取图片路径没有后缀名问题，修复没有设置裁剪获取图片路径为null的问题
* ios fix ios 13 get image path without suffix

## 1.0.6+1
* ios修复ios13获取相册路径错误问题
* iOS fix iOS13 getting album path error problem
* android优化预览多图片时内存问题
* Android optimization memory issues when previewing multiple pictures
## 1.0.6
* 添加预览多图功能
* Add preview multi-picture function
* 修改ui主题颜色功能，由原来的预制ui颜色改为动态设置ui颜色
* Modify the ui theme color function from the original pre-made ui color to dynamically set the ui color
* 修复无法展示gif图片功能（注意：忽略了gif裁剪/压缩功能）
* Fix unable to display gif image function (note: gif crop/compress function is ignored)
## 1.0.5+3
* 安卓修复加载库报错问题
* Android fixes some mobile phone loading library error
## 1.0.5+2
* 修复项目主页重定向问题
* Fix project homepage redirection problem
## 1.0.5+1
* Modify bug
## 1.0.5
* 适配安卓安卓Q，增加ui主题配置功能，增加直接拍照和录制视频功能
* Adapt Android Android Q, add ui theme configuration function, increase direct photo and record video function
## 1.0.4+1
* Fix Android P can't preview image
## 1.0.4
* configuration description
## 1.0.3
* Increase the save network video to album function
## 1.0.2
* Support Android iOS
* Android i OS supports local language environment (Chinese or English)
* image_pickers Support picture selection, video multiple selection, support to save network pictures to albums, support preview video and preview picture function


android/src/main/java/com/leeson/image_pickers/activitys/SelectPicsActivity.java
package com.leeson.image_pickers.activitys;

import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.media.ThumbnailUtils;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;

import com.leeson.image_pickers.AppPath;
import com.leeson.image_pickers.R;
import com.leeson.image_pickers.utils.CommonUtils;
import com.leeson.image_pickers.utils.GlideEngine;
import com.leeson.image_pickers.utils.PictureStyleUtil;
import com.luck.picture.lib.PictureSelectionModel;
import com.luck.picture.lib.PictureSelector;
import com.luck.picture.lib.camera.CustomCameraView;
import com.luck.picture.lib.config.PictureConfig;
import com.luck.picture.lib.config.PictureMimeType;
import com.luck.picture.lib.entity.LocalMedia;
import com.luck.picture.lib.tools.AndroidQTransformUtils;
import com.luck.picture.lib.tools.PictureFileUtils;
import com.luck.picture.lib.tools.SdkVersionUtils;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import top.zibin.luban.CompressionPredicate;
import top.zibin.luban.Luban;
import top.zibin.luban.OnCompressListener;
import top.zibin.luban.OnRenameListener;


/**
 * Created by lisen on 2018-09-11.
 * 只选择多张图片，
 *
 * @author lisen < 453354858@qq.com >
 */
@SuppressWarnings("all")
public class SelectPicsActivity extends BaseActivity {

    private static final int WRITE_SDCARD = 101;

    public static final String GALLERY_MODE = "GALLERY_MODE";
    public static final String UI_COLOR = "UI_COLOR";
    public static final String SHOW_GIF = "SHOW_GIF";
    public static final String SHOW_CAMERA = "SHOW_CAMERA";
    public static final String ENABLE_CROP = "ENABLE_CROP";
    public static final String WIDTH = "WIDTH";
    public static final String HEIGHT = "HEIGHT";
    public static final String COMPRESS_SIZE = "COMPRESS_SIZE";

    public static final String SELECT_COUNT = "SELECT_COUNT";//可选择的数量

    public static final String COMPRESS_PATHS = "COMPRESS_PATHS";//压缩的画
    public static final String CAMERA_MIME_TYPE = "CAMERA_MIME_TYPE";//直接调用拍照或拍视频时有效
    private Number compressSize;
    private int compressCount = 0;
    private String mode;
    private Map<String,Number> uiColor;
    private Number selectCount;
    private boolean showGif;
    private boolean showCamera;
    private boolean enableCrop;
    private Number width;
    private Number height;
    private String mimeType;

    @Override
    public void onCreate(@androidx.annotation.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_select_pics);
        mode = getIntent().getStringExtra(GALLERY_MODE);
        uiColor = (Map<String, Number>) getIntent().getSerializableExtra(UI_COLOR);

        selectCount = getIntent().getIntExtra(SELECT_COUNT, 9);
        showGif = getIntent().getBooleanExtra(SHOW_GIF, true);
        showCamera = getIntent().getBooleanExtra(SHOW_CAMERA, false);
        enableCrop = getIntent().getBooleanExtra(ENABLE_CROP, false);
        width = getIntent().getIntExtra(WIDTH, 1);
        height = getIntent().getIntExtra(HEIGHT, 1);
        compressSize = getIntent().getIntExtra(COMPRESS_SIZE, 500);
        mimeType = getIntent().getStringExtra(CAMERA_MIME_TYPE);

        startSel();
    }

    private void startSel(){
        PictureStyleUtil pictureStyleUtil = new PictureStyleUtil(this);

        //添加图片
        PictureSelector pictureSelector = PictureSelector.create(this);
        PictureSelectionModel pictureSelectionModel = null;
        if (mimeType != null){
            //直接调用拍照或拍视频时
            if ("photo".equals(mimeType)) {
                pictureSelectionModel = pictureSelector.openCamera(PictureMimeType.ofImage());
                pictureSelectionModel.setButtonFeatures(CustomCameraView.BUTTON_STATE_ONLY_CAPTURE);
                if (SdkVersionUtils.checkedAndroid_Q()){
                    pictureSelectionModel.imageFormat(PictureMimeType.PNG_Q);
                }else{
                    pictureSelectionModel.imageFormat(PictureMimeType.PNG);
                }
                //安卓端自定义相机拍照有问题，暂时不用了，等原项目修复bug
                pictureSelectionModel.isUseCustomCamera(false);
            } else {
                pictureSelectionModel = pictureSelector.openCamera(PictureMimeType.ofVideo());
                pictureSelectionModel.setButtonFeatures(CustomCameraView.BUTTON_STATE_ONLY_RECORDER);
                pictureSelectionModel.imageFormat(PictureMimeType.MIME_TYPE_VIDEO);
                pictureSelectionModel.recordVideoSecond(10);
                pictureSelectionModel.isUseCustomCamera(true);
            }
        }else{
            //从相册中选择
            if ("image".equals(mode)){
                pictureSelectionModel = pictureSelector.openGallery(PictureMimeType.ofImage());
                pictureSelectionModel.setButtonFeatures(CustomCameraView.BUTTON_STATE_ONLY_CAPTURE);
                if (SdkVersionUtils.checkedAndroid_Q()){
                    pictureSelectionModel.imageFormat(PictureMimeType.PNG_Q);
                }else{
                    pictureSelectionModel.imageFormat(PictureMimeType.PNG);
                }
                pictureSelectionModel.isUseCustomCamera(true);
            }else{
                pictureSelectionModel = pictureSelector.openGallery(PictureMimeType.ofVideo());
                pictureSelectionModel.setButtonFeatures(CustomCameraView.BUTTON_STATE_ONLY_RECORDER);
                pictureSelectionModel.imageFormat(PictureMimeType.MIME_TYPE_VIDEO);
//                pictureSelectionModel.videoMaxSecond(10);//相册展示的视频最大时长
                pictureSelectionModel.recordVideoSecond(10);
                pictureSelectionModel.isUseCustomCamera(true);
            }
        }

        pictureSelectionModel
                .loadImageEngine(GlideEngine.createGlideEngine())
                .isOpenStyleNumComplete(true)
                .isOpenStyleCheckNumMode(true)

//                .imageFormat(
//                        SdkVersionUtils.checkedAndroid_Q()
//                        ? ("video".equals(mimeType)  ? PictureMimeType.MIME_TYPE_VIDEO  : PictureMimeType.PNG_Q.toLowerCase() )
//                        : ("video".equals(mimeType) ? PictureMimeType.MIME_TYPE_VIDEO : PictureMimeType.PNG.toLowerCase()))

                .setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT)
                .setPictureStyle(pictureStyleUtil.getStyle(uiColor))
                .setPictureCropStyle(pictureStyleUtil.getCropStyle(uiColor))

//                .imageFormat(PictureMimeType.PNG.toLowerCase())// 拍照保存图片格式后缀,默认jpeg
                .isCamera(showCamera)
                .isGif(showGif)
                .maxSelectNum(enableCrop?1:selectCount.intValue())
                .withAspectRatio(width.intValue(), height.intValue())
                .imageSpanCount(4)// 每行显示个数 int
                .selectionMode(enableCrop || selectCount.intValue() == 1 ? PictureConfig.SINGLE : PictureConfig.MULTIPLE)// 多选 or 单选 PictureConfig.MULTIPLE or PictureConfig.SINGLE
                .isSingleDirectReturn(true)// 单选模式下是否直接返回
                .previewImage(true)// 是否可预览图片 true or false
                .enableCrop(enableCrop)// 是否裁剪 true or false
                .isWeChatStyle(true)
                .circleDimmedLayer(false)
                .showCropFrame(true)
                .showCropGrid(true)
                .hideBottomControls(true)
                .freeStyleCropEnabled(false)

                .compress(false)// 是否压缩 true or false
                .minimumCompressSize(Integer.MAX_VALUE)
                .compressSavePath(getPath())//压缩图片保存地址
                .forResult(PictureConfig.CHOOSE_REQUEST);

    }
    private String getPath() {
        String path = new AppPath(this).getAppImgDirPath(false);
        File file = new File(path);
        if (file.mkdirs()) {
            createNomedia(path);
            return path;
        }
        createNomedia(path);
        return path;
    }

    private void createNomedia(String path) {
        File nomedia = new File(path, ".nomedia");
        if (!nomedia.exists()) {
            try {
                nomedia.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case PictureConfig.CHOOSE_REQUEST:
                    // 图片、视频、音频选择结果回调
                    List<LocalMedia> selectList = PictureSelector.obtainMultipleResult(data);

                    List<String> paths = new ArrayList<>();
                    for (int i = 0; i < selectList.size(); i++) {
                        LocalMedia localMedia = selectList.get(i);
                        if (localMedia.isCut()) {//2.5.9 android Q 裁剪没问题
                            // 因为这个lib中 gif裁剪有问题，所以gif裁剪过就不使用裁剪地址，使用原gif地址
                            if (Build.VERSION.SDK_INT >= 29) {
                                if (localMedia.getPath() != null && localMedia.getAndroidQToPath() != null && localMedia.getAndroidQToPath().endsWith(".gif")){
                                    String path = PictureFileUtils.getPath(getApplicationContext(), Uri.parse(localMedia.getPath()));
                                    paths.add(path);
                                }else{
                                    paths.add(localMedia.getCutPath());
                                }
                            } else {
                                if (localMedia.getPath() != null && localMedia.getPath().endsWith(".gif")){
                                    paths.add(localMedia.getPath());
                                }else{
                                    paths.add(localMedia.getCutPath());
                                }
                            }
                        } else {
                            if (Build.VERSION.SDK_INT >= 29) {
                                //图片选择库 2.5.9 有bug，要这样处理
                                String AndroidQToPath = AndroidQTransformUtils.copyPathToAndroidQ(SelectPicsActivity.this,System.currentTimeMillis(),
                                        localMedia.getPath(), localMedia.getWidth(), localMedia.getHeight(), localMedia.getMimeType(), localMedia.getRealPath().substring(localMedia.getRealPath().lastIndexOf("/")+1));
                                localMedia.setAndroidQToPath(AndroidQToPath);

                                paths.add(localMedia.getAndroidQToPath());
                            } else {
                                paths.add(localMedia.getPath());
                            }
                        }
                    }

                    if (mimeType != null){
                        //直接调用拍照或拍视频时
                        if ("photo".equals(mimeType)) {
                            lubanCompress(paths);
                        }else{
                            resolveVideoPath(paths);
                        }
                    }else{
                        if ("image".equals(mode)) {
                            //如果选择的是图片就压缩
                            lubanCompress(paths);
                        } else {
                            resolveVideoPath(paths);
                        }
                    }
                    break;
                case WRITE_SDCARD:


                    break;
            }
        } else {
            finish();
        }
    }


    private void resolveVideoPath(List<String> paths) {

        List<Map<String, String>> thumbPaths = new ArrayList<>();
        for (int i = 0; i < paths.size(); i++) {
            String path = paths.get(i);
            Bitmap bitmap = ThumbnailUtils.createVideoThumbnail(path, MediaStore.Video.Thumbnails.FULL_SCREEN_KIND);
            String thumbPath = CommonUtils.saveBitmap(this, new AppPath(this).getAppImgDirPath(false), bitmap);
            Map<String, String> map = new HashMap<>();
            map.put("thumbPath", thumbPath);
            map.put("path", path);
            thumbPaths.add(map);
        }
        Intent intent = new Intent();
        intent.putExtra(COMPRESS_PATHS, (Serializable) thumbPaths);
        setResult(RESULT_OK, intent);
        finish();
    }

    private void lubanCompress(final List<String> paths) {
        final List<Map<String, String>> lubanCompressPaths = new ArrayList<>();
        Luban.with(this)
                .load(paths)
                .ignoreBy(compressSize.intValue())
                .setTargetDir(getPath())
                .filter(new CompressionPredicate() {
                    @Override
                    public boolean apply(String path) {
                        return !path.endsWith(".gif");
                    }
                })
                .setRenameListener(new OnRenameListener() {
                    @Override
                    public String rename(String filePath) {
                        return filePath.substring(filePath.lastIndexOf("/"));
                    }
                })
                .setCompressListener(new OnCompressListener() {
                    @Override
                    public void onStart() {
                    }

                    @Override
                    public void onSuccess(File file) {
                        // 压缩成功后调用，返回压缩后的图片文件
                        Map<String, String> map = new HashMap<>();
                        map.put("thumbPath", file.getAbsolutePath());
                        map.put("path", file.getAbsolutePath());
                        lubanCompressPaths.add(map);
                        compressCount++;
                        compressFinish(paths, lubanCompressPaths);
                    }

                    @Override
                    public void onError(Throwable e) {
                        // 当压缩过程出现问题时调用
                        compressCount++;
                        compressFinish(paths, lubanCompressPaths);
                    }
                }).launch();
    }

    private void compressFinish(List<String> paths, List<Map<String, String>> compressPaths) {
        if (compressCount == paths.size()) {
            Intent intent = new Intent();
            intent.putExtra(COMPRESS_PATHS, (Serializable) compressPaths);
            setResult(RESULT_OK, intent);
            finish();
        }
    }
}


lib/image_pickers.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/cupertino.dart';
import 'package:flutter/services.dart';

enum GalleryMode {
  ///选择图片
  image,

  ///选择视频
  video,
}

enum CameraMimeType {
  ///拍照
  photo,

  ///拍视频
  video,
}

class ImagePickers {
  static const MethodChannel _channel =
      const MethodChannel('flutter/image_pickers');

  /// 返回拍摄的图片或视频的信息 Return information of the selected picture or video
  ///
  /// cameraMimeType CameraMimeType.photo为拍照，CameraMimeType.video 为录制视频 CameraMimeType.photo is a photo, CameraMimeType.video is a video
  ///
  ///cropConfig 裁剪配置（视频不支持裁剪和压缩，当选择视频时此参数无效） Crop configuration (video does not support cropping and compression, this parameter is not available when selecting video)
  ///
  ///compressSize 拍照后（录制视频时此参数无效）的忽略压缩大小，当图片大小小于compressSize时将不压缩 单位 KB Ignore compression size after selection, will not compress unit KB when the image size is smaller than compressSize
  ///
  static Future<Media> openCamera({
    CameraMimeType cameraMimeType: CameraMimeType.photo,
    CropConfig? cropConfig,
    int compressSize: 500,
  }) async {
    String mimeType = "photo";
    if (cameraMimeType == CameraMimeType.video) {
      mimeType = "video";
    }

    bool enableCrop = false;
    int width = 1;
    int height = 1;
    if (cropConfig != null) {
      enableCrop = cropConfig.enableCrop;
      width = cropConfig.width <= 0 ? 1 : cropConfig.width;
      height = cropConfig.height <= 0 ? 1 : cropConfig.height;
    }

    Color uiColor = UIConfig.defUiThemeColor;
    final Map<String, dynamic> params = <String, dynamic>{
      'galleryMode': "image",
      'showGif': true,
      'uiColor': {
        "a": 255,
        "r": uiColor.red,
        "g": uiColor.green,
        "b": uiColor.blue,
        "l": (uiColor.computeLuminance() * 255).toInt()
      },
      'selectCount': 1,
      'showCamera': false,
      'enableCrop': enableCrop,
      'width': width,
      'height': height,
      'compressSize': compressSize < 50 ? 50 : compressSize,
      'cameraMimeType': mimeType,
    };
    final List<dynamic>? paths =
        await _channel.invokeMethod('getPickerPaths', params);

    if (paths != null && paths.length > 0) {
      Media media = Media();
      media.thumbPath = paths[0]["thumbPath"];
      media.path = paths[0]["path"];
      media.galleryMode = GalleryMode.image;
      return media;
    }

    return Media();
  }

  ///选择图片或视频 Choose an image or video
  ///
  ///返回选择的图片或视频的信息 Return information of the selected picture or video
  ///
  ///
  ///galleryMode 选择图片或者选择视频 枚举 Select an image or select a video to enumerate
  ///
  /// uiConfig 选择图片或选择视频页面的主题 默认 0xfffefefe Select an image or select the theme of the video page Default 0xfffefefe
  ///
  ///selectCount 要选择的图片数量 Number of images to select
  ///
  ///showCamera 是否显示相机按钮 Whether to display the camera button
  ///
  ///cropConfig 裁剪配置（视频不支持裁剪和压缩，当选择视频时此参数无效） Crop configuration (video does not support cropping and compression, this parameter is not available when selecting video)
  ///
  ///compressSize 选择图片（选择视频时此参数无效）后的忽略压缩大小，当图片大小小于compressSize时将不压缩 单位 KB Ignore compression size after selection, will not compress unit KB when the image size is smaller than compressSize
  ///

  static Future<List<Media>> pickerPaths({
    GalleryMode galleryMode: GalleryMode.image,
    UIConfig? uiConfig,
    int selectCount: 1,
    bool showCamera: false,
    bool showGif: true,
    CropConfig? cropConfig,
    int compressSize: 500,
  }) async {
    String gMode = "image";
    if (galleryMode == GalleryMode.image) {
      gMode = "image";
    } else if (galleryMode == GalleryMode.video) {
      gMode = "video";
    }
    Color uiColor = UIConfig.defUiThemeColor;
    if (uiConfig != null) {
      uiColor = uiConfig.uiThemeColor;
    }

    bool enableCrop = false;
    int width = 1;
    int height = 1;
    if (cropConfig != null) {
      enableCrop = cropConfig.enableCrop;
      width = cropConfig.width <= 0 ? 1 : cropConfig.width;
      height = cropConfig.height <= 0 ? 1 : cropConfig.height;
    }

    final Map<String, dynamic> params = <String, dynamic>{
      'galleryMode': gMode,
      'showGif': showGif,
      'uiColor': {
        "a": 255,
        "r": uiColor.red,
        "g": uiColor.green,
        "b": uiColor.blue,
        "l": (uiColor.computeLuminance() * 255).toInt()
      },
      'selectCount': selectCount,
      'showCamera': showCamera,
      'enableCrop': enableCrop,
      'width': width,
      'height': height,
      'compressSize': compressSize < 50 ? 50 : compressSize,
    };
    final List<dynamic> paths = await _channel.invokeMethod('getPickerPaths', params);
    List<Media> medias = [];
    paths.forEach((data) {
      Media media = Media();
      media.thumbPath = data["thumbPath"];
      media.path = data["path"];
      media.galleryMode = galleryMode;
      medias.add(media);
    });
    return medias;
  }

  ///预览图片 preview picture
  ///
  ///imagePath 图片本地路径或者网络url Image local path or web url

  static previewImage(String imagePath) {
    final Map<String, dynamic> params = <String, dynamic>{
      'path': imagePath,
    };
    _channel.invokeMethod('previewImage', params);
  }

  ///预览多张图片 Preview multiple pictures
  ///
  ///imagePaths 图片本地路径集合或者网络url集合 Image local path collection or network url collection

  static previewImages(List<String?> imagePaths, int initIndex) {
    final Map<String, dynamic> params = <String, dynamic>{
      'paths': imagePaths,
      'initIndex': initIndex,
    };
    _channel.invokeMethod('previewImages', params);
  }

  ///预览多张图片 Preview multiple pictures
  ///
  ///imageMedias 图片数据集合 Image local path collection or network url collection
  ///
  ///Media中真正有效使用的数据是 Media.path The really effectively used data in Media is Media.path
  ///
  static previewImagesByMedia(List<Media> imageMedias, int initIndex) {
    if (imageMedias.length > 0) {
      List<String?> paths =
          imageMedias.map((Media media) => media.path).toList();
      previewImages(paths, initIndex);
    }
  }

  ///预览视频 Preview video
  ///
  /// videoPath 图片本地路径或者网络url Image local path or web url
  ///
  /// thumbPath 视频封面图 Video cover

  static previewVideo(String videoPath, {String thumbPath: ""}) {
    final Map<String, dynamic> params = <String, dynamic>{
      'path': videoPath,
      'thumbPath': thumbPath,
    };
    _channel.invokeMethod('previewVideo', params);
  }

  ///保存图片字节到相册中 Save picture bytes to album
  ///
  /// data 图片字节数据 Picture byte data
  ///

  static Future<String?> saveByteDataImageToGallery(
    Uint8List data,
  ) async {
    final Map<String, dynamic> params = <String, dynamic>{
      'uint8List': data,
    };
    String? path =
        await _channel.invokeMethod('saveByteDataImageToGallery', params);
    return path;
  }

  ///保存图片到相册中 Save image to album
  ///
  /// 返回保存的图片路径 Return the saved image path
  ///
  /// imageUrl 网络图片url Web image url

  static Future<String?> saveImageToGallery(String imageUrl) async {
    final Map<String, dynamic> params = <String, dynamic>{
      'path': imageUrl,
    };
    String? path = await _channel.invokeMethod('saveImageToGallery', params);
    return path;
  }

  ///保存视频到相册中 Save video to album
  ///
  /// 返回保存的视频路径 Android 返回保存的视频文件路径 ios 暂无法返回本地视频路径
  /// Return the saved video path Android returns the saved video file path ios can not return to the local video path
  ///
  /// videoUrl 网络视频url Web video url
  ///
  static Future<String?> saveVideoToGallery(String videoUrl) async {
    final Map<String, dynamic> params = <String, dynamic>{
      'path': videoUrl,
    };
    String? path = await _channel.invokeMethod('saveVideoToGallery', params);
    return path;
  }
}

///裁剪配置
///Crop configuration
class CropConfig {
  ///是否可裁剪
  bool enableCrop = false;

  ///裁剪的宽度比例
  ///Cropped width ratio
  int width = 1;

  ///裁剪的高度比例
  ///Crop height ratio
  int height = 1;

  CropConfig({this.enableCrop: false, this.width: 1, this.height: 1});
}

class Media {
  ///视频缩略图图片路径
  ///Video thumbnail image path
  String? thumbPath;

  ///视频路径或图片路径
  ///Video path or image path
  String? path;
  GalleryMode? galleryMode;
}

/// Created by liSen on 2019/11/15 10:51.
///
/// @author liSen < 453354858@qq.com >
///
/// 选择图片页面颜色配置
///
/// Select image page color configuration
///
class UIConfig {
  static const Color defUiThemeColor = Color(0xfffefefe);
  Color uiThemeColor;

  /// uiThemeColor
  UIConfig({this.uiThemeColor: defUiThemeColor});
}
